### DISCUSSION TITLE
Can someone explain when exactly implicit transactions occur?

### DISCUSSION THREAD

**Discussion by ersinakinci** [mikro-orm/mikro-orm#5466] [Q&A] [answered]
I understand the "unit of work" model, or at least I think I do, but the docs aren't clear on how different API methods map onto transactions.

For example, if I want to create new DB records, I can do

```
const myRecord = new MyEntity(...);
em.persist(myRecord);

const anotherRecord = new AnotherEntity(...);
em.persist(anotherRecord);

em.flush();  // Writes both records to the db within a transaction
```

But I can also do

```
await em.create(MyEntity, { ... });
await em.create(AnotherEntity, { ... });

em.flush();  // Is this even necessary? Does it do anything?
```

It's unclear to me that the second approach results in records being created within an implicit transaction.

Likewise, using `.find` seems to instantly retrieve records--are those statements wrapped inside of a transaction?

**Comment by B4nan** [comment, marked as answer]
[`em.create`](https://mikro-orm.io/api/core/class/EntityManager#create) is sync method, it does the same as your previous example - create entity instance (via constructor) and persist it. its the flush call that does things (and you need to await that one instead of the create call). that's described [here](https://mikro-orm.io/docs/guide/relationships#creating-entity-graph)

> Likewise, using .find seems to instantly retrieve records--are those statements wrapped inside of a transaction?

Nope, you need explicit transaction demarcation for that. 

https://mikro-orm.io/docs/transactions#approach-2-explicitly

With that approach, you no longer need the explicit flush, as it is called automatically after the callback (as well as via explicit `em.commit()`).

**Reply by ersinakinci** [reply]
Thank you @B4nan! Also didn't realize that `em.create` was sync, so all these `await`s I'm using are unnecessary.

What happens if you do a `.find` with an em instance that's waiting to be flushed? Does it do the query "out of band," so to speak, and allow the other persisted transactions to be flushed later on as usual? Or does it flush right away?

**Reply by B4nan** [reply]
There is a queue for the flush calls (or better say `uow.commit` calls), they need to got sequentially as they work with the state. But you can fork and flush forks in parallel if that's what you want.

edit: but the `find` call will be resolved right ahead, that's not blocking on its own

**Reply by ersinakinci** [reply]
Great, thanks!
